### PR TITLE
[Azure] Add NAT gateways for outbound public IP addresses

### PR DESCRIPTION
**Pull request by seankimkdy**
Takes on Azure portion of #434. This allows resources to have outbound permit list rules with public IP addresses. When a permit list rule with a public IP address target is passed in to `AddPermitListRules`, a NAT gateway is spun up in the vnet where the resource resides. Similar to VPN gateways, the NAT gateway is created on demand and not before it is necessary.

> [!NOTE]  
> Initiating inbound connections is still not supported and can only be done if a user attaches a public IP address to the VM. Support for this may come later but hasn't been discussed.

I also refactored Azure's connectivity check methods such that it's more user friendly to those from other clouds.

GCP PR is coming soon after #417 is merged due to a lot of code overlap.

### CODE REVIEW

File: pkg/azure/plugin.go [hunk 1]
```diff
@@ -173,10 +173,49 @@ func (s *azurePluginServer) AddPermitListRules(ctx context.Context, req *paragli
 
 		for i, peeringCloudInfo := range peeringCloudInfos {
 			if peeringCloudInfo == nil {
-				continue
-			}
-			address := rule.Targets[i]
-			if peeringCloudInfo.Cloud != utils.AZURE {
+				// Setup NAT gateway for public IP addresses
+				natGatewayLocation := *resourceVnet.Location
```
Comment (smcclure20) [inline pkg/azure/plugin.go:177]: Nit: Maybe this should be its own function?

File: pkg/azure/test_utils.go [hunk 2]
```diff
@@ -754,3 +779,123 @@ func getFakeVnetInLocation(location *string, addressSpace string) *armnetwork.Vi
...
+		if *resp.ConnectivityInformation.ConnectionStatus == armnetwork.ConnectionStatus("Reachable") {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+func RunICMPConnectivityCheck(ctx context.Context, namespace string, subscriptionId string, resourceGroupName string, sourceVmName string, destinationIPAddress string, tries int) (bool, error) {
+	// Note that external ICMP
```
Comment (smcclure20) [inline pkg/azure/test_utils.go:889]: Unfinished comment?

**Review by smcclure20 — CHANGES_REQUESTED** [review]
Just a couple nits! Thanks!

File: pkg/azure/test_utils.go (hunk 2 above)
Comment (seankimkdy) [reply pkg/azure/test_utils.go]: Good catch!

**Review by smcclure20 — APPROVED** [review]

**Pull request merged**